### DISCUSSION TITLE
Interoperability of stateright with Zenoh (or other similar pub/sub middlewares)?

### DISCUSSION THREAD

**Discussion by andrew-otiv** [stateright/stateright#72] [Q&A]
First of all, thanks for sharing such an interesting library with the world!

We're building a remote control system for the rail industry. We're planning to build our protocols for control and telemetry of the train on top of a pub/sub middleware (i.e. zenoh or MQTT). We have a few node types for different subsystems on and off of the locomotives, each of which will have its own state machine.

Is stateright and its Actor model something that would be suitable for implementing state machines communicating over Zenoh, or is stateright currently tied to its internal communications protocol?

If it's not advisable to hook a pub/sub middleware into stateright, is it feasible to do model checking using stateright's message passing code, but share the state machine logic with the middleware-using production implementation? 

(Note that the notion of "production" is a bit complicated in our industry since eventually we will require some level of functional safety certification, but it's a long development journey to get there)

How far could we go with checking an implementation without compiling stateright into the production code, since doing so would require stateright itself to also have functional safety certification? i.e. could we conditionally implement traits that get us model checking but only switch them on in debug builds?

Sorry for the barrage of open ended questions; we're just getting started with formal verification and model checking, and figuring how they could fit in our development and testing campaign. 

(oeps, response below is also me; accidentally used my personal account)

**Comment by drewm1980** [comment]
I read the book, studied the relevant examples, and thought some more!

Among possible first steps, I think modeling generic pub/sub on top of stateright would be a good one, and is probably enough to expose performance or ergonomics issues with pub/sub.

Design decisions:
* Model a single message broker that models the middleware in a separate Actor, similar in spirit to brokered middlewares like MQTT or model the middleware with code in each actor, i.e. similar to peer to peer topology possible in Zenoh?
* Model channel subscription at initialization time, statically, dynamically, or both?
* Model discovery at initialization time, statically, dynamically, or both?

This would be fun to take a crack at; maybe I'll work on it as an exercise, time permitting. If you already know precedents for this in other model checking frameworks similar to stateright that would also be interesting!
